### ISSUE TITLE
Out of Memory errors with internal resolutions > 1x

### ISSUE THREAD

**Issue by rlyerly** [Themaister/parallel-rdp#19]
Hello,

I'm using ParaLLEl RDP in Retroarch on a system with the following specs:

* Windows 10
* CPU: Intel 8705G with Intel HD 630 iGPU and 8GB RAM
* GPU: Radeon Vega M GL with 4GB HBM (distinct memory from CPU RAM)
* Radeon Adrenalin 20.4.1 drivers, last "official" driver for this GPU
* Retroarch v1.9.0
* Core: Mupen64Plus

-Next

I can select either GPU via Windows graphics settings.  When I use the iGPU, everything works as expected - I'm able to run with internal resolutions above 1x, although framerates drop at 4x because the iGPU is fairly weak.  When I use the Vega M GL at 1x internal resolution, it also works correctly.  However whenever I set the internal resolution to >1x and launch the emulator, Retroarch opens a window and hangs forever.  If I click on the window, Windows says it's not responding and prompts me to force close the emulator.

Here's the Retroarch log for one of these runs, although nothing seems out of the ordinary minus the hang:
[retroarch-log.txt](https://github.com/Themaister/parallel-rdp/files/5699669/retroarch-log.txt)

I also managed to catch the console log for when it hangs:
[console-output.txt](https://github.com/Themaister/parallel-rdp/files/5699670/console-output.txt).  It looks like submitting commands the queue for the GPU context fails with out of memory errors (according to the [VkResult](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkResult.html) documentation).

I don't really expect ParaLLEl RDP to use 4 GB of memory but maybe I'm wrong.  Are there any memory requirements I'm missing?

**Comment by rlyerly** [comment]
I did some digging and the GPU is definitely not running out of memory, I added print statements and it was only allocating several hundred MBs of GPU RAM.  Instead I think my AMD Vulkan driver only allows allocating a puny number of buffers in a descriptor set (maxDescriptorSetStorageBuffersDynamic is 8 according to [the Vulkan database](https://vulkan.gpuinfo.org/displayreport.php?id=9992#properties)).  I started seeing the out of memory errors shown in the log above during the 9th call to `Renderer::create_buffer` (inside `Renderer::init_buffers`).

It looks like [several additional buffers are allocated](https://github.com/Themaister/parallel-rdp/blob/master/parallel-rdp/rdp_renderer.cpp#L284-L315) if ubershaders are not enabled.  I tried force-enabling ubershaders using environment variables, but it still hits those errors because `Renderer::init_caps()` (where ubershaders is initialized via environment variables) is called after `Renderer::init_buffers()`.  However after recompiling with [ubershader default initialized to true](https://github.com/Themaister/parallel-rdp/blob/master/parallel-rdp/rdp_renderer.hpp#L354), I'm able to get in game.

**Comment by Themaister** [comment]
parallel-rdp does not use dynamic SSBOs, so that doesn't make much sense.

**Comment by Themaister** [comment]
An other user had it working just fine on an AMD APU with 4x and even 8x upscaling (albeit horribly slow as expected), so I don't know.

**Comment by rlyerly** [comment]
Yeah, unfortunately my system is kind of a weird one - it's one of the Kaby Lake G laptops with the Vega M GPU that Intel was supposed to support, then AMD kind of took over support, then neither wanted anything to do with.  The driver I have is the last officially supported one, so it looks like the bugs are here to stay.

You're also right about my previous theory being wrong - when I bumped the resolution up to 4x, it throws similar out of memory errors on [when creating upscaling_multisampled_hidden_rdram in rdp_renderer.cpp](https://github.com/Themaister/parallel-rdp/blob/master/parallel-rdp/rdp_renderer.cpp#L575).

If I go back to the original configuration (ubershaders default initialized to false, 2x internal resolution), I see errors starting earlier when creating [multisampled_rdram](https://github.com/Themaister/parallel-rdp/blob/master/parallel-rdp/rdp_renderer.cpp#L571).

I'm going to try setting up some of the Vulkan validation layers to see if that turns up anything.  I'm pretty sure I'm not running out of GPU memory, because I printed the values from [`get_memory_budget()`](https://github.com/Themaister/parallel-rdp-standalone/blob/master/vulkan/device.hpp#L322) for all heaps as each allocation was happening and there was plenty of free memory in all of them.  There's got to be some other resource that is getting exhausted.
